### PR TITLE
Properly unwrap partial async functions to direct them to the correct async scheduler.

### PR DESCRIPTION
**Pull request by shipmints**
Per #397

### CODE REVIEW

**Comment by agronholm** [comment]
This PR does not pass the test suite. Please ensure that it does. I will leave a review pointing out other problems with this. It should also be noted that explicit Tornado support will be removed in v4.0 because Tornado is asyncio compatible.

File: apscheduler/executors/asyncio.py [hunk 1]
```diff
@@ -1,8 +1,8 @@
-from asyncio import iscoroutinefunction
 import sys
 
 from apscheduler.executors.base import BaseExecutor, run_job, run_coroutine_job
 
+from apscheduler.util import iscoroutinefunction_partial
```
Comment (agronholm) [inline apscheduler/executors/asyncio.py]: This belongs in the same import group as the previous import.

File: apscheduler/executors/tornado.py [hunk 2]
```diff
@@ -1,11 +1,12 @@
 import sys
 from concurrent.futures import ThreadPoolExecutor
-from inspect import iscoroutinefunction
 
 from tornado.gen import convert_yielded
 
 from apscheduler.executors.base import BaseExecutor, run_job, run_coroutine_job
 
+from apscheduler.util import iscoroutinefunction_partial
```
Comment (agronholm) [inline apscheduler/executors/tornado.py]: Ditto.

File: apscheduler/util.py [hunk 3]
```diff
@@ -4,7 +4,7 @@
 from calendar import timegm
 from inspect import signature
 from functools import partial
-from inspect import isclass, ismethod
+from inspect import isclass, ismethod, iscoroutinefunction
```
Comment (agronholm) [inline apscheduler/util.py]: You need to use `asyncio.iscoroutinefunction()` here so it catches coroutines created in asyncio debug mode.

File: apscheduler/util.py [hunk 4]
```diff
@@ -384,3 +384,9 @@ def check_callable_args(func, args, kwargs):
         raise ValueError(
             'The target callable does not accept the following keyword arguments: %s' %
             ', '.join(unmatched_kwargs))
+
+
+def iscoroutinefunction_partial(f):
+    while isinstance(f, partial):
+        f = f.func
+    return iscoroutinefunction(f)
```
Comment (agronholm) [inline apscheduler/util.py]: Add a blank line after the `while` block.

File: tests/test_util.py [hunk 5]
```diff
@@ -323,3 +323,22 @@ def test_unfulfilled_kwargs(self):
...
+    pass
+
+
+async def a_coro(x):
+    pass
+
+
+class TestIsCoroutineFunctionPartial(object):
+    def test_non_coro(self):
+        assert iscoroutinefunction_partial(not_a_coro) is False
```
Comment (agronholm) [inline tests/test_util.py]: Replace with `assert not ...` as per PEP 8.

File: tests/test_util.py [hunk 6]
```diff
@@ -323,3 +323,22 @@ def test_unfulfilled_kwargs(self):
         exc = pytest.raises(ValueError, check_callable_args, func, [1], {})
         assert str(exc.value) == ('The following keyword-only arguments have not been supplied in '
                                   'kwargs: y')
+
+
+def not_a_coro(x):
```
Comment (agronholm) [inline tests/test_util.py]: These should probably go inside the new test class.

File: tests/test_util.py [hunk 7]
```diff
@@ -323,3 +323,22 @@ def test_unfulfilled_kwargs(self):
...
+async def a_coro(x):
+    pass
+
+
+class TestIsCoroutineFunctionPartial(object):
+    def test_non_coro(self):
+        assert iscoroutinefunction_partial(not_a_coro) is False
+
+    def test_coro(self):
+        assert iscoroutinefunction_partial(a_coro) is True
```
Comment (agronholm) [inline tests/test_util.py]: Just remove the `is True` part.

File: tests/test_util.py [hunk 8]
```diff
@@ -323,3 +323,22 @@ def test_unfulfilled_kwargs(self):
...
+
+    def test_coro_partial(self):
+        assert iscoroutinefunction_partial(partial(a_coro, 1)) is True
```
Comment (agronholm) [inline tests/test_util.py]: Ditto.

File: apscheduler/executors/asyncio.py (hunk 1 above)
Comment (shipmints) [reply apscheduler/executors/asyncio.py]: I will ensure these abut.

File: apscheduler/executors/tornado.py (hunk 2 above)
Comment (shipmints) [reply apscheduler/executors/tornado.py]: I will ensure these abut.

File: apscheduler/util.py (hunk 3 above)
Comment (shipmints) [reply apscheduler/util.py]: I've explicitly made it say asyncio.iscoroutinefunction in situ so readers know precisely which version is being called. Not clear this is about debug mode so much as decorator-based async functions ala @coroutine.

File: apscheduler/util.py (hunk 4 above)
Comment (shipmints) [reply apscheduler/util.py]: These will no longer abut.

File: tests/test_util.py (hunk 5 above)
Comment (shipmints) [reply tests/test_util.py]: Amended.

File: tests/test_util.py (hunk 6 above)
Comment (shipmints) [reply tests/test_util.py]: Amended.

File: tests/test_util.py (hunk 7 above)
Comment (shipmints) [reply tests/test_util.py]: Amended.

File: tests/test_util.py (hunk 8 above)
Comment (shipmints) [reply tests/test_util.py]: Amended. Review commended. Thank you.

**Comment by agronholm** [comment]
I am in process of refactoring this for the 3.x branch since v4.0 will not have separate asyncio/tornado executors.

**Comment by agronholm** [comment]
Took a while but now the code is compatible with the 3.x branch.

**Comment by agronholm** [comment]
Thanks!

**Pull request merged**